torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -215,15 +216,232 @@ class LocalSyncInserter {
+          // Inline allocation, place after expr
+          place_after = expr;
+        } else {
+          // Place allocation after the last computeAt axis
+          // TODO: may be more efficient to place after the first non-computeAt
+          // axis
+          place_after = *(loops_it + 1);
+        }
+
+        place_in->body().insert_after(place_after, sync_expr);
+      }
+    }
+  }
+
+  void visit(kir::ForLoop* fl) final {
+    for_loops.push_back(fl);
+    // Modifying in place, make a copy of the vector
+    const std::vector<kir::Expr*> exprs = fl->body().exprs();
+    for (auto expr : exprs) {
+      handle(expr);
+    }
+    for_loops.pop_back();
+  }
+
+  void visit(kir::IfThenElse*) final {
+    TORCH_INTERNAL_ASSERT(
+        false,
+        "Pass does not support conditional statements, ",
+        "this pass should be run before any conditionals are placed in code.");
+  }
+
+  // Clear the modify status for all shared memory buffers
+  static void cleanSharedMemory(std::unordered_map<kir::Val*, bool>& smem) {
+    for (auto& item : smem) {
+      item.second = false;
+    }
+  }
+
+  // Return the status of the shared memory buffer
+  // False if TensorView is not shared memory buffer
+  bool isModifiedSharedMemory(
+      std::unordered_map<kir::Val*, bool>& smem,
+      std::vector<kir::Val*> keys) const {
+    return std::any_of(keys.begin(), keys.end(), [&smem](kir::Val* key) {
+      auto it = smem.find(key);
+      if (it != smem.end()) {
+        return it->second;
+      }
+      return false;
+    });
+  }
+
+  ReadAfterWriteSyncs(std::vector<kir::Expr*> _loop_nests)
+      : loop_nests_(_loop_nests) {
+    // Fusion shared_memory values
+    // Tracks if shared memory is modified
+    std::unordered_map<kir::Val*, bool> smem;
+
+    // Flatten all the expressions
+    auto flattened_exprs = ExprFlattener::flatten(loop_nests_);
+
+    kir::Expr* prev_tv_expr = nullptr;
+    for (auto expr : flattened_exprs) {
+      if (!ir_utils::isTVOp(expr) || expr->isA<kir::Allocate>()) {
+        continue;
+      }
+
+      bool need_sync = isModifiedSharedMemory(smem, expr->inputs());
+      if (need_sync) {
+        TORCH_INTERNAL_ASSERT(
+            prev_tv_expr != nullptr,
+            "Can't require sync on inputs, however, detected it's needed.");
+        sync_after.push_back(prev_tv_expr);
+        cleanSharedMemory(smem);
+      }
+
+      for (auto out : expr->outputs()) {
+        if (out->isA<kir::TensorView>()) {
+          if (out->as<kir::TensorView>()->memoryType() == MemoryType::Shared) {
+            smem[out] = true;
+          }
+        }
+      }
+
+      prev_tv_expr = expr;
+    }
+
+    // Insert read after write syncs
+    const std::vector<kir::Expr*> exprs = loop_nests_;
+    for (auto expr : exprs) {
+      handle(expr);
+    }
+
+    TORCH_INTERNAL_ASSERT(
+        sync_after.empty(), "Didn't place all required syncs.");
+  }
+
+ private:
+  std::deque<kir::Expr*> sync_after;
```

Comment:
naming: _ suffix

also, can you please add a brief comment for each of these data memembers?